design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+
+
+
+### Milestone 2: Update K8s secret with Conjur secret
+
+#### Design
+
+Once a secret has been updated in Conjur, this will trigger an update in K8s Secrets values. For Milestone 2, the Secrets Provider deployment will need to be a Pod that does not terminate (instead of a Job). Every change in Conjur secret, will trigger an event to update the K8s Secret.
+
+|      | Trigger for Secret Rotation                                  | Elaboration                                                  | Pros                                                         | Cons                                                         |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | *Every 5 minutes,* retrieve secrets every constant, configurable amount of time. Possible options:<br /><br />a. Run constantly and wake up when time is right <br />b. Spin-up every 5 minutes, run using [cron job](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) based on K8s secrets update (Get trigger upon secret change) | *Flow*<br />1. Every 5 minutes, Secrets Provider checks if a change/a new K8s Secret was created <br />2. If so, update Secrets Provider local cache of state for K8s | - Easily integrate with current solution since retry mechanism already exists<br />- No need to change anything in Conjur | -  Polling is a bad way for changes notification mechanism.<br />- Option (a) is inefficient management of resources.<br />- Option (b) is supported only since K8s version 1.18 and is currently in beta. |
+| 2    | *Outside trigger,* register to 3rd party pub/sub message handler to be notified of secrets change.<br />Ex: <br />- ZMQ (no agent)<br />- RabbitMQ |                                                              | - Do the job only when really needed - very efficient<br />- Immediate update of K8s Secret on secret change in Conjur<br />- Integrated option for registration only on relevant changes | - Need development in Conjur side to update the 3rd party component with changes<br />- Requires another component to deploy and maintain (if not ZMQ) |
+| 3    | *Outside trigger,* retrieve secrets when notified from Follower by webhook registered API | *Concern*<br />Follower does not save state and cannot save the webhook URL | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Follower may not be able to reach secrets provider (need to save URL in Master Database)<br />- Significant development required because it requires a new event mechanism on server |
```

Comment:
Followers do have state as we replicate the master state into the follower. so any state the master have, the follower can have it as well